src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,176 @@
+//    - principal_1 canActAs principal_2       (a canActAs)
+// To encapsulate which form this takes, it:
+//    - uses factory methods to make either form
+//    - uses a private std::variant member that takes either of these two forms
+//    - uses std::visit to do operations on the actual state
+class VerbPhrase {
+ public:
+  explicit VerbPhrase(Predicate pred) : internal_form_(std::move(pred)) {}
+  explicit VerbPhrase(Principal right_prin)
+      : internal_form_(std::move(right_prin)) {}
+
+ private:
+  // The principal variant of the internal form means this represents
+  // "canActAs principal".
+  std::variant<Predicate, Principal> internal_form_;
+};
+
+// This the AST Node for FlatFact which takes on one of two forms:
+//   - <principal> <verb_phrase>  (* a principal applied to a verbphrase *)
+//   - <predicate>                (* just a predicate *)
+// A similar variant-based pattern is used for this as for VerbPhrase. Here an
+// inner struct is used for the first form (principal applied to verb_phrase).
+class FlatFact {
+ public:
+  explicit FlatFact(Principal prin, VerbPhrase verb_phrase)
+      : internal_form_(
+            VerbPhraseFlatFact(std::move(prin), std::move(verb_phrase))) {}
+  explicit FlatFact(Predicate pred) : internal_form_(std::move(pred)) {}
+
+ private:
+  struct VerbPhraseFlatFact {
+    Principal prin;
+    VerbPhrase verb_phrase;
+    VerbPhraseFlatFact(Principal prin, VerbPhrase verb_phrase)
+        : prin(std::move(prin)), verb_phrase(std::move(verb_phrase)) {}
+  };
+  std::variant<VerbPhraseFlatFact, Predicate> internal_form_;
+};
+
+// This is the AST node for facts which have two forms:
+//  - <flat_fact>
+//  - <principal> canSay <flat_fact>
+class Fact {
+ public:
+  explicit Fact(FlatFact flat_fact) : internal_form_(std::move(flat_fact)) {}
+  explicit Fact(Principal prin, Fact* fact)
+      : internal_form_(CanSayFact(std::move(prin), fact)) {}
+
+ private:
+  struct CanSayFact {
+    Principal prin;
+    Fact* fact;
+    CanSayFact(Principal prin, Fact* fact)
+        : prin(std::move(prin)), fact(fact) {}
+  };
+  std::variant<FlatFact, CanSayFact> internal_form_;
+};
+
+// AST Node for Assertion (this is a sub-expression of SaysAssertion)
+// which has two forms:
+//  - <fact>                                      (unconditional)
+//  - <fact> :- <flat_fact_1> ... <flat_fact_n>   (conditional)
+class Assertion {
+ public:
+  explicit Assertion(Fact fact) : internal_form_(std::move(fact)) {}
+  explicit Assertion(Fact lhs, std::vector<FlatFact> rhs)
+      : internal_form_(CondAssertion(std::move(lhs), std::move(rhs))) {}
+
+ private:
+  struct CondAssertion {
+    Fact lhs;
+    std::vector<FlatFact> rhs;
+    CondAssertion(Fact lhs, std::vector<FlatFact> rhs)
+        : lhs(std::move(lhs)), rhs(std::move(rhs)) {}
+  };
+  std::variant<Fact, CondAssertion> internal_form_;
+};
+
+// AST node for saysAssertion, which prepends an assertion with:
+// "<principal> says"
+class SaysAssertion {
```

Comment:
Do we generally want to group all assertions made by a `Principal`? If so, would it be better to add a map to the `Program` data structure below?

Reply:
I think there is not a use-case for any compiler passes the Rust verison has (so we won't need them here either). If we need this later, I thing we'd be better off doing it in a separate pass (and implementing this just when we need it).